proxy/proxy.go

```diff
@@ -115,630 +116,275 @@ func (p *Server) isStopped() bool {
-		// Skip connection-specific headers
-		if strings.ToLower(name) == "connection" || strings.ToLower(name) == "proxy-connection" {
-			continue
-		}
-		for _, value := range values {
-			req.Header.Add(name, value)
+		err := conn.Close()
+		if err != nil {
+			p.logger.Error("Failed to close connection", "error", err)
 		}
-	}
-
-	p.logger.Debug("About to make HTTP request", "target", targetURL.String())
-	resp, err := client.Do(req)
-	if err != nil {
-		p.logger.Error("Failed to make forward request", "error", err, "target", targetURL.String(), "error_type", fmt.Sprintf("%T", err))
-		http.Error(w, fmt.Sprintf("Failed to make request: %v", err), http.StatusBadGateway)
 		return
 	}
-	defer func() { _ = resp.Body.Close() }()
-
-	p.logger.Debug("Received response", "status", resp.StatusCode, "target", targetURL.String())
-
-	// Copy response headers (except connection-specific ones)
-	for name, values := range resp.Header {
-		if strings.ToLower(name) == "connection" || strings.ToLower(name) == "transfer-encoding" {
-			continue
-		}
-		for _, value := range values {
-			w.Header().Add(name, value)
-		}
-	}
-
-	// Copy status code
-	w.WriteHeader(resp.StatusCode)
-
-	// Copy response body
-	bytesWritten, copyErr := io.Copy(w, resp.Body)
-	if copyErr != nil {
-		p.logger.Error("Error copying response body", "error", copyErr, "bytes_written", bytesWritten)
-		http.Error(w, "Failed to copy response", http.StatusBadGateway)
+	if isTLS {
+		p.logger.Debug("🔒 Detected TLS connection - handling as HTTPS")
+		p.handleTLSConnection(wrappedConn)
 	} else {
-		p.logger.Debug("Successfully forwarded HTTP response", "bytes_written", bytesWritten, "status", resp.StatusCode)
-	}
-
-	// Ensure response is flushed
-	if flusher, ok := w.(http.Flusher); ok {
-		flusher.Flush()
+		p.logger.Debug("🌐 Detected HTTP connection")
+		p.handleHTTPConnection(wrappedConn)
 	}
-	p.logger.Debug("forwardHTTPRequest completed")
 }
 
-// writeBlockedResponse writes a blocked response
-func (p *Server) writeBlockedResponse(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "text/plain")
-	w.WriteHeader(http.StatusForbidden)
-
-	// Extract host from URL for cleaner display
-	host := r.URL.Host
-	if host == "" {
-		host = r.Host
+func (p *Server) isTLSConnection(conn net.Conn) (net.Conn, bool, error) {
+	// Read first byte to detect TLS
+	buf := make([]byte, 1)
+	n, err := conn.Read(buf)
+	if err != nil || n == 0 {
+		return nil, false, fmt.Errorf("failed to read first byte from connection: %v, read %v bytes", err, n)
 	}
 
-	_, _ = fmt.Fprintf(w, `🚫 Request Blocked by Boundary
-
-Request: %s %s
-Host: %s
-
-To allow this request, restart boundary with:
-  --allow "%s"                    # Allow all methods to this host
-  --allow "%s %s"          # Allow only %s requests to this host
-
-For more help: https://github.com/coder/boundary
-`,
-		r.Method, r.URL.Path, host, host, r.Method, host, r.Method)
-}
+	connWrapper := &connectionWrapper{conn, buf, false}
```

Comment:
imo using buffered reader peak would be nicer than this connectionWrapper type: https://pkg.go.dev/bufio#Reader

Reply:
I'll take a look into that. I kept this part from proxy-v1 implementation.
Proxy-v1 had many bugs, but connectionWrapper worked fine.